### PR TITLE
Update README

### PR DESCRIPTION
**Pull request by KDr2**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -3,29 +3,128 @@
 Orignally, this package has some benchmarking scripts for Turing. Now,
```
Comment (trappmartin) [inline README.md]: Change to: "This package evolved from Turing.jl specific benchmarking tools to a generic benchmarking library for continuous benchmarking."

File: README.md [hunk 2]
```diff
@@ -3,29 +3,128 @@
...
 it evolves into a generic benchmarking library for Julia packages.
 
-It can run in two modes:
+It can run in many modes:
```
Comment (trappmartin) [inline README.md]: It can be used in the following scenarios:

File: README.md [hunk 3]
```diff
@@ -3,29 +3,128 @@
...
-and a repository for storing the benchmark report.
+ContinuousBenchmarks doesn't offer tools to measure metrics of
+interests, so to benchmark your code, you should resort tools like the
+Julia built-in macros for benchmarks, like `@timev`, `@timed`,
+`@elapsed`, and `@allocated`, or the package `BenchmarkTools.jl`.
 
-If you want it to run as a GitHub App, a GitHub App should also be set
-up, an its information, like app id, pem, etc., should be put into the
-configuration file as well.
+We use a simple protocol to collect measured metrics from the
```
Comment (trappmartin) [inline README.md]: Is `LOG_DATA` a global variable? Is the same variable used in all tests? And if so why don't we need to do something like:
```julia
LOG_DATA = merge(LOG_DATA, Dict("time" => t, .... ))
``` 

I might misunderstand something here.

File: README.md [hunk 4]
```diff
@@ -3,29 +3,128 @@
...
+benchmarks: storing the measured metrics in a global variable named
+`LOG_DATA` which is either a `Dict` or a `DataFrame`, for example:
 
-If you only want to run it on Travis-CI, just fork this repository,
-and update the `config/app.travis.toml` file.
+```
```
Comment (trappmartin) [inline README.md]: Maybe use 
```
```julia
```

instead.

File: README.md [hunk 5]
```diff
@@ -3,29 +3,128 @@
...
+# file: benchmark1.jl
+val, t, bytes, gctime, memallocs = @timed rand(10^6)
+LOG_DATA = Dict("time" => t, "memory" => bytes, "gctime" => gctime)
```
Comment (trappmartin) [inline README.md]: Is the dict a `Dict{String, Any}`? Would your code still work if the user would do something like this?

```julia
val, t, bytes, gctime, memallocs = @timed rand(10^6)
pt = plot(rand(100))
LOG_DATA = Dict("time" => t, "memory" => bytes, "gctime" => gctime, "trace plot" => pt)
```

File: README.md [hunk 6]
```diff
@@ -3,29 +3,128 @@
...
+
+# or
+# LOG_DATA = DataFrame(time=[t], memory=[bytes], gctime=[gctime])
+```
+
+You can report more than one records by using the `DataFrame`
```
Comment (trappmartin) [inline README.md]: - typo "one record" not "records"
- remove "by"

File: README.md [hunk 7]
```diff
@@ -3,29 +3,128 @@
...
+type. It's worth noting that all the records reported from benchmarks
```
Comment (trappmartin) [inline README.md]: "Note that all records reported during continuous benchmarking should report using the metrics as Dict keys and DataFrame colums are shared."

I think this quite a restriction. Would it possible to relax this? Why is this necessary at the first place?

File: README.md [hunk 8]
```diff
@@ -3,29 +3,128 @@
...
+  benchmark, the benchmarks will run in a Julia process started by
+  `julia --project=<target.project_dir>`.
+- `benchmark_files` is a list whose elements are the absolute path of
+  the benchmark files.
+
+### Run benchmarks on Travis-CI
+
+When the benchmarks and the config file are ready, we can write a
+script to run them on Travis, and here is an example:
+https://github.com/TuringLang/Turing.jl/blob/master/benchmarks/runbenchmarks.jl
```
Comment (trappmartin) [inline README.md]: This is super cool!

File: README.md [hunk 9]
```diff
@@ -3,29 +3,128 @@
...
+.
+
+When someone pushes new commits to the repository, the script will run
```
Comment (trappmartin) [inline README.md]: Is it possible to only run the benchmarks if a commit invokes this by having `#benchmark` in the commit message?

Example:

```
commit message: init --> no benchmark
commit message: added code --> no benchmark
commit message: added test #benchmark --> run benchmark
commit message: bug fix --> no benchmark
commit message: improvements #benchmark --> run benchmark
```

File: README.md [hunk 10]
```diff
@@ -3,29 +3,128 @@
...
+separately:
+
+```
+using ContinuousBenchmarks
+using ContinuousBenchmarks.Runner
+ContinuousBenchmarks.set_benchmark_config_file("/path/to/config.jl")
+Runner.run_bm_on_travis("$BM_JOB_NAME", ("$BASE_BRANCH", "$CURRENT_BRANCH"), "$COMMIT_SHA")
+```
+
+After it completes, all the measured metrics will be collected, then a
```
Comment (trappmartin) [inline README.md]: If the benchmark contains plots, would we store them too? I think this might be quite useful.

File: README.md [hunk 11]
```diff
@@ -3,29 +3,128 @@
...
+Then we put the GitHub information in the config file and start the
+app server:
+
+```
+export TBB_CONFIG=/path/to/config/file # TOML one
+julia -e 'using ContinuousBenchmarks; ContinuousBenchmarks.AppServer.main()'
+```
 
-Here how we run a GitHub App for Turing.jl:
+Here how we use a GitHub App for Turing.jl:
```
Comment (trappmartin) [inline README.md]: "Here we use the Github App for Turing.jl as an example:"

File: README.md [hunk 12]
```diff
@@ -49,7 +148,7 @@ Here how we run a GitHub App for Turing.jl:
   (https://github.com/TuringBenchBot) and a TuringBenchBot GitHub App
   (https://github.com/apps/turingbenchbot);
 - When you want to do a benchmark, `@` the GitHub Account in a comment,
-  tell it the target branches, e.g. `@TuringBenchBot bm("master",
+  tell it the target branches, e.g. `@BayesianBot bm("master",
```
Comment (trappmartin) [inline README.md]: So it will only be invoked if the user calls `@TuringBenchBot bm("branch1", "branch2")`? Cool!

File: README.md [hunk 13]
```diff
@@ -63,23 +162,13 @@ Here how we run a GitHub App for Turing.jl:
...
-ContinuousBenchmarks.set_project_path(".")
-report_path = Runner.local_benchmark("TEST", ("master", "new_branch"))
-```
+ContinuousBenchmarks uses a `toml` file as its default configuration
+file, there's a template at `config/app.toml.tmpl`.
 
-3. Open the `report.md` report file at `report_path` to view
-   benchmarking results. More details can be found in the subdirectory
-   of each commit which are in the same path as the report.
+When the benchmark job starts, the `toml` config file according to the
```
Comment (trappmartin) [inline README.md]: Could you rewrite this? I'm not sure I understand what is happening.

You load the `toml` file specified in the environment variable `TBB_CONFIG` and then?

File: README.md (hunk 3 above)
Comment (KDr2) [reply README.md]: Each benchmark runs in its own Julia process, so each benchmark has its own `LOG_DATA` variable, though it is global. 

I added this line in.

File: README.md (hunk 5 above)
Comment (KDr2) [reply README.md]: The plot will be converted to a string by calling `repr` or `show`.

I will create a new issue to support rich format data in the report.

File: README.md (hunk 7 above)
Comment (KDr2) [reply README.md]: Now, I put all of the collected data in one table, so the records should better have the same columns. 

We sure can relax this, but we'd better find a way to show them clearly in the report.

File: README.md (hunk 9 above)
Comment (KDr2) [reply README.md]: Sure, actually I did this for Turing.jl, just add these lines to the file https://github.com/TuringLang/Turing.jl/blob/master/benchmarks/runbenchmarks.jl:

```julia
COMMIT_MSG = strip(read(`git show -s --format="%s"`, String))
occursin("[bm]", COMMIT_MSG) || exit(0)
```

But yebai said we'd better run the benchmarks by default. If one wants to skip the benchmarks,  he can add `[skip ci]` to the commit msg to skip all the CI jobs, including the benchmarks. So I removed these lines.

File: README.md (hunk 10 above)
Comment (KDr2) [reply README.md]: We can do this, I will create a new issue for this.

File: README.md (hunk 12 above)
Comment (KDr2) [reply README.md]: For the GitHub App, Yes :)

File: README.md (hunk 1 above)
Comment (KDr2) [reply README.md]: done

File: README.md (hunk 2 above)
Comment (KDr2) [reply README.md]: Done.

File: README.md (hunk 11 above)
Comment (KDr2) [reply README.md]: done.

File: README.md (hunk 13 above)
Comment (KDr2) [reply README.md]: update.

File: README.md (hunk 10 above)
Comment (KDr2) [reply README.md]: https://github.com/TuringLang/ContinuousBenchmarks.jl/issues/51

File: README.md (hunk 9 above)
Comment (trappmartin) [reply README.md]: @yebai what is the reason for running the benchmarks by default? Because we will only have small benchmarks, e.g. MvNormal, tested by Travis-CI?

File: README.md (hunk 7 above)
Comment (trappmartin) [reply README.md]: Is the issue that the report will be a single table for all benchmark runs we did?

I think it would be the best if you can report different metrics for different benchmarks. I don't see why the should depend on each other. You should be able to do something like:

benchmark1.jl
```
LOG_DATA = Dict("time" => t, "memory" => bytes, "gctime" => gctime)
```

benchmark2.jl
```
LOG_DATA = Dict("time" => t, "ess" => ess, "rhat" => rhat, "m" => mean(m), "histm" => histogram(m) )
```

and so on.

And the resulting markdown would create a table for each benchmark comparing two branches.

Does that make sense?

Comment (KDr2) [reply README.md]: It makes sense,  I create a new issue for this: https://github.com/TuringLang/ContinuousBenchmarks.jl/issues/52

File: README.md (hunk 9 above)
Comment (yebai) [reply README.md]: Yeah, Travis-CI is only meant for small benchmarks. It's trying to have a lightweight benchmarking for all PRs in the future, so we can inspect the performance changes before merging.
